### PR TITLE
[dev] DeepSeek-V4 CSA optimization: fuse some THD-related ops

### PR DESCRIPTION
**Pull request by hxbai**
- [ ] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?
Reduce CUDA launch and temporary-tensor overhead in fused CSA packed-THD training by fusing layout lowering, position mapping, indexer preprocessing, and sparse-loss postprocessing.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/transformer/experimental_attention_variant/csa_utils/fused_compressor.py [hunk 1]
```diff
@@ -143,7 +147,8 @@ def fused_compressor_available(device: Optional[torch.device] = None) -> bool:
         return False
     supported = _DEVICE_SUPPORTED_CACHE.get(index)
     if supported is None:
-        supported = torch.cuda.get_device_capability(index) == _SUPPORTED_COMPUTE_CAPABILITY
+        capability = torch.cuda.get_device_capability(index)
+        supported = capability[0] >= _MINIMUM_COMPUTE_CAPABILITY_MAJOR
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/experimental_attention_variant/csa_utils/fused_compressor.py:151]: It is intentional. won't fix it.

File: megatron/core/transformer/experimental_attention_variant/csa_utils/csa_indexer_loss_kernels.py [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+    if padding_row_mask is not None:
+        row_mask = padding_row_mask.unsqueeze(-1)
+        sanitized_topk_indices = topk_indices.masked_fill(row_mask, -1)
+        if physical_indices is not None:
+            sanitized_physical_indices = physical_indices.masked_fill(row_mask, -1)
+
+    safe_indices = sanitized_topk_indices.clamp(min=0).long()
+    gathered_scores = torch.gather(indexer_scores, dim=-1, index=safe_indices)
+    gathered_scores = torch.where(sanitized_topk_indices >= 0, gathered_scores, _FLOAT32_MIN)
+    predict = torch.softmax(gathered_scores, dim=-1)
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/experimental_attention_variant/csa_utils/csa_indexer_loss_kernels.py]: fixed

File: experimental/lite/megatron/lite/primitive/modules/attention/csa.py [hunk 3]
```diff
@@ -872,25 +879,16 @@ def _forward_thd_cp(
...
-                )
+            (
+                hidden_compact,
+                compressed_group_ids,
+                compressed_position_ids,
+                _local_cu_seqlens,
+                _local_cu_seqlens_compressed,
+                cu_seqlens_compressed,
+                seq_to_rank_row,
+            ) = cp_utils.prepare_cp_compressor_input(
```
[the remark this answers is not in the document]
Comment (hxbai) [reply experimental/lite/megatron/lite/primitive/modules/attention/csa.py:887]: added the TODO

File: megatron/core/transformer/experimental_attention_variant/csa_utils/cp_layout_kernels.py [hunk 4]
```diff
@@ -1,840 +1,28 @@
...
+build_attention_indices = _impl.build_attention_indices
 
-try:
-    import cuda.bindings.driver as cuda
-    import cutlass
-    import cutlass.cute as cute
-    from cutlass import utils
-    from cutlass.cute.runtime import from_dlpack, make_fake_stream
+# Preserve the availability probe used by downstream tests and diagnostics.
+_CUTE_AVAILABLE = _impl._CUTE_AVAILABLE
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/experimental_attention_variant/csa_utils/cp_layout_kernels.py]: fixed

File: megatron/core/transformer/experimental_attention_variant/csa_utils/thd_layout_kernels.py [hunk 5]
```diff
@@ -0,0 +1,1152 @@
...
+
+        # Metadata is consumed by a later launch on the same stream. Keeping the
+        # short prefix scans in this launch avoids the eager diff/div/cumsum/cat
+        # chains without requiring grid-wide synchronization.
+        if bidx == 0 and tidx == 0:
+            local_running_groups = 0
+            global_running_groups = 0
+            local_cu_seqlens[0] = 0
+            local_cu_seqlens_comp[0] = 0
+            cu_seqlens_comp[0] = 0
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/experimental_attention_variant/csa_utils/thd_layout_kernels.py:110]: fixed

File: tests/unit_tests/transformer/experimental_attention_variant/test_dsv4_hybrid_attention_cp.py [hunk 6]
```diff
@@ -843,10 +843,16 @@ def test_thd_cp_zero_indexer_loss_keeps_indexer_grads(self):
         del attn, full_hidden, full_grad, hidden, grad
         _clear_cuda_test_state()
 
-    @pytest.mark.parametrize("apply_rope_fusion", [True, False], ids=["fused", "unfused"])
-    def test_dynamic_cp_mla_up_proj_recompute_matches_eager(self, apply_rope_fusion):
+    @pytest.mark.parametrize(
+        "use_fused_kernels,apply_rope_fusion",
+        [(False, True), (False, False), (True, True)],
+        ids=["unfused-attn-fused-rope", "unfused-attn-unfused-rope", "fused-attn"],
+    )
```
[the remark this answers is not in the document]
Comment (hxbai) [reply tests/unit_tests/transformer/experimental_attention_variant/test_dsv4_hybrid_attention_cp.py:960]: fixed

File: megatron/core/fusions/fused_mla_yarn_rope_apply.py [hunk 7]
```diff
@@ -66,6 +66,24 @@ def _get_thd_token_idx(cu_seqlens, pid_m, seq_num, cp_rank, cp_size):
...
+    """Map a row in a contiguous THD slice to its position within the packed sequence."""
+    global_row = pid_m.to(tl.int64) + global_start
+    token_idx = tl.full((), 0, dtype=tl.int64)
+    seq_idx = 0
+    while seq_idx < seq_num:
+        seq_start = tl.load(cu_seqlens + seq_idx).to(tl.int64)
+        seq_end = tl.load(cu_seqlens + seq_idx + 1).to(tl.int64)
+        in_sequence = (global_row >= seq_start) & (global_row < seq_end)
+        token_idx = tl.where(in_sequence, global_row - seq_start, token_idx)
+        seq_idx += 1
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/fusions/fused_mla_yarn_rope_apply.py]: changed to bin search for better perf

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/7064/commits/46041a306c5071357dca0263f43e93a121ff69fb

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/7064/commits/555e462d53f802da382662d3201850d3519ba41a

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/7064/commits/66e0fee63ae36a4a19a792584cd0d9d0847bc11c

**Review by yaox12 — APPROVED** [review]
